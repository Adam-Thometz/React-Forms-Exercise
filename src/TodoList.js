import React, {useState} from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";

const TodoList = () => {
  const [todos, setTodos] = useState([])

  const addTodo = ({id, todo}) => {
    setTodos(todos => [...todos, {id, todo}])
  }

  const removeTodo = id => {
    setTodos(todos => todos.filter(t => t.id !== id))
  }

  const editTodo = (id, edited) => {
    setTodos(todos => todos.map(
      todo => todo.id === id ? {...todo, todo: edited} : todo
    ))
  }

  return (
    <div className="TodoList">
      <NewTodoForm handleAdd={addTodo} />
      {todos.map(t =>
        <Todo
          key={t.id}
          id={t.id}
          todo={t.todo}
          remove={removeTodo}
          edit={editTodo} />)}
    </div>
  )
}

export default TodoList