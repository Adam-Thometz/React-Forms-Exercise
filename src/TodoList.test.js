import { render, fireEvent, waitFor } from '@testing-library/react';
import TodoList from './TodoList';

const addTodo = async (todoList) => {
  const todoInput = todoList.getByLabelText("New Todo")
  const submitBtn = todoList.getByText("Create Todo")

  await waitFor(() => {
    fireEvent.change(todoInput, {target: {value: "buy a bucket"}})
    fireEvent.click(submitBtn)
  })
}

it('renders without crashing', () => {
  render(<TodoList />);
});

it('matches snapshot', () => {
  const {asFragment} = render(<TodoList />)
  expect(asFragment()).toMatchSnapshot()
});

it("can add a new todo", async () => {
  const todoList = render(<TodoList />)
  await addTodo(todoList)

  expect(todoList.getByText("buy a bucket")).toBeInTheDocument()
  expect(todoList.getByText("Edit")).toBeInTheDocument()
  expect(todoList.getByText("Remove")).toBeInTheDocument()
})

it("can remove a todo", async () => {
  const todoList = render(<TodoList />)
  await addTodo(todoList)

  const removeBtn = todoList.getByText("Remove")
  fireEvent.click(removeBtn)
  expect(removeBtn).not.toBeInTheDocument()
})

it("can edit a new todo", async () => {
  const todoList = render(<TodoList />)
  await addTodo(todoList)
  
  const editBtn = todoList.getByText("Edit")
  fireEvent.click(editBtn)

  const editInput = todoList.getByDisplayValue("buy a bucket")
  await waitFor(() => {
    fireEvent.change(editInput, {target: {value: 'write tests'}})
    fireEvent.click(todoList.getByText('Update'))
  })
  
  expect(todoList.getByText('write tests')).toBeInTheDocument()
  expect(todoList.getByText('buy a bucket')).not.toBeInTheDocument()
})