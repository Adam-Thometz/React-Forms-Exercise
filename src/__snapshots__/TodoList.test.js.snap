// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <form
      class="NewTodoForm"
    >
      <label
        for="todo"
      >
        New Todo
      </label>
      <input
        id="todo"
        name="todo"
        placeholder="Enter todo here"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Create Todo
      </button>
    </form>
  </div>
</DocumentFragment>
`;
